[![Python 3.7](https://img.shields.io/badge/python-3.7-blue.svg)](https://www.python.org/downloads/release/python-370/)



# Troika

Detection of resistance mechanisms in _Mycobacterium tuberculosis_ is dependent upon identification of SNPs that may confer decreased susceptibility to anti-mycobacterial drugs. TBrnr is a pipeline, which implements [TB-profiler](https://github.com/jodyphelan/TBProfiler) and its related database to detect resistance conferring mutations from Illumina read data and filters these results for reporting for public health and clinical use in Australia.


### Motivation

There are many tools and databases available, however, for the purposes of reporting genomic AST for _M. tuberculosis_ in the context of public health and clinical use in Australia customisation is required. Rather than reinventing the wheel, TBrnr leverages a high quality database and detection tool which is highly customisable.


## Pipeline

TBrnr is designed for batch reporting of AST in _M. tuberculosis_ isolates generated from Illumina reads.

### Installation

TO COME

#### Conda (Recomended)

#### PyPi


### Running TBrnr

TO COME

